kyo-data/shared/src/main/scala/kyo/Tag.scala

```diff
@@ -1,239 +1,541 @@
+            end apply
+        end UnionType
+
+        /** Represents the variance of a type parameter. Variance affects subtyping relationships for parameterized types.
+          *
+          * @param toString
+          *   String representation of the variance
+          */
+        enum Variance(override val toString: String) extends Serializable derives CanEqual:
+            /** Invariant parameters require exact type matches */
+            case Invariant extends Variance("")
+
+            /** Covariant parameters allow subtypes (+) */
+            case Covariant extends Variance("+")
+
+            /** Contravariant parameters allow supertypes (-) */
+            case Contravariant extends Variance("-")
+        end Variance
+
+        /** Represents a class or trait type, potentially with type parameters. This includes all named types in Scala like List, Option,
+          * etc.
+          *
+          * @param id
+          *   A unique identifier for this class type
+          * @param bases
+          *   The base classes/traits of this type, including itself
+          * @tparam A
+          *   The Scala type represented by this class type
+          */
+        final case class ClassType[A](
+            id: String,
+            bases: Chunk.Indexed[ClassType.Base]
+        ) extends Type[A]:
+            def className: String                  = bases(0).className
+            def variances: Chunk.Indexed[Variance] = bases(0).variances
+            def params: Chunk.Indexed[Type[?]]     = bases(0).params
+
+            override def toString =
+                val str = variances.zip(params).map((v, p) => s"$v$p").mkString(", ")
+                val p =
+                    if str.isBlank() then ""
+                    else s"[$str]"
+                s"$className$p"
+            end toString
+        end ClassType
+
+        object ClassType:
+            /** Represents a base class or trait of a class type. This includes information about the class name, type parameter variances,
+              * and the actual type arguments.
+              *
+              * @param className
+              *   The fully qualified name of the class
+              * @param variances
+              *   The variances of the type parameters
+              * @param params
+              *   The actual type arguments
+              */
+            final case class Base(
+                className: String,
+                variances: Chunk.Indexed[Variance],
+                params: Chunk.Indexed[Type[?]]
+            )
+        end ClassType
+
+        /** Represents a recursive type reference. This is used to handle recursive types without infinite recursion.
+          *
+          * @param id
+          *   The identifier of the referenced type
+          * @tparam A
+          *   The Scala type represented by this recursive reference
+          */
+        case class Recursive[A](id: String) extends Type[A]
```

Comment:
This is a case we didn't have before because `Tag` was ignoring type parameters of super classes. An example case is `String` since it extends from `Comparable[String]`, which would recurse the type in the AST.